### PR TITLE
[SPARK-9082][SQL] Filter using non-deterministic expressions should not be pushed down

### PR DESCRIPTION
**Pull request by cloud-fan**

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @yhuai

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/FilterPushdownSuite.scala [hunk 1]
```diff
@@ -146,6 +146,32 @@ class FilterPushdownSuite extends PlanTest {
     comparePlans(optimized, correctAnswer)
   }
 
+  test("nondeterministic: can't push down filter through project") {
+    val originalQuery = testRelation
+      .select(Rand(10).as('rand))
+      .where('rand > 5)
```
Comment (yhuai) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/FilterPushdownSuite.scala]: Maybe it is better to use a condition having both deterministic and non-deterministic expressions, e.g.
```
 val originalQuery = testRelation
  .select(Rand(10).as('rand), 'a)
  .where('rand > 5 && 'a > 5)
```

**Comment by yhuai** [comment]
Thanks for the fix! There is one thing I think it is better to double check. Let's say I have a partitioned Parquet table and I select a few columns from it and then I have predicates on top of it. These predicates have non-deterministic expressions, a predicate on the partition column, and some predicates that can be pushed down to parquet table scan (pushed into parquet's reader). With out fix, will partitioning column get correctly pruned and those predicates get pushed down to parquet's reader? `ParquetFilterSuite.scala` is the file for testing parquet filter pushdown. We may need to add tests for partition pruning. @liancheng Where are our tests for partition pruning?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/FilterPushdownSuite.scala [hunk 2]
```diff
@@ -146,6 +146,32 @@ class FilterPushdownSuite extends PlanTest {
     comparePlans(optimized, correctAnswer)
   }
 
+  test("nondeterministic: can't push down filter through project") {
+    val originalQuery = testRelation
+      .select(Rand(10).as('rand), 'a)
+      .where('rand > 5 && 'a > 5)
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/FilterPushdownSuite.scala]: looks like we should optimize in to
```scala
testRelation
  .where('a > 5)
  .select(Rand(10).as('rand), 'a)
  .where('rand > 5)
```
Can we do this?

Comment (yhuai) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/FilterPushdownSuite.scala]: I feel it may require much more changes. I am fine if we do that in a separate pr in future.

**Comment by liancheng** [comment]
@yhuai I believe expressions with UDF(s) are never pushed down in Parquet or any other data sources. Only simple comparison and string predicates dealing with constants can be pushed down. I'm double checking partition pruning.

**Comment by yhuai** [comment]
Yeah. But my concern is if this fix will prevent any legitimate predicates from being pushed down.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 3]
```diff
@@ -513,18 +513,28 @@ object SimplifyFilters extends Rule[LogicalPlan] {
...
+      })
 
-  private def replaceAlias(condition: Expression, sourceAliases: Map[Attribute, Expression]) = {
-    condition transform {
-      case a: AttributeReference => sourceAliases.getOrElse(a, a)
-    }
+      // We only push down filter if their overlapped expressions are all
+      // deterministic.
+      val hasNondeterministic = condition.collect {
+        case a: Attribute if aliasMap.contains(a) => aliasMap(a)
```
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: We probably can not use the `contains` here, as we have to use the `semanticEquals` for finding the identical expression.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 4]
```diff
@@ -513,18 +513,28 @@ object SimplifyFilters extends Rule[LogicalPlan] {
...
+      }.exists(_.find(!_.deterministic).isDefined)
+
+      if (hasNondeterministic) {
+        filter
```
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: We still can push down the deterministic part. For example:
```
testRelation
  .where('a > 5)
  .select(Rand(10).as('rand), 'a)
  .where('rand > 5 and 'a > 8)
```
==>
```
testRelation
  .where('a > 5)
  .where('a>8)
  .select(Rand(10).as('rand), 'a)
  .where('rand > 5)
```

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: yea, I also mentioned it in a comment for the test. Will do it in a follow-up PR.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 3 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: good catch!

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: But `aliasMap` is `AttributeMap`, I think it should be safe to call `contains` here?

Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Oh, yes, the original code seems has bug, it use the `.toMap`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 4 above)
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: oh, sorry, I didn't notice the comments in the unit test. But I think we'd better to do that in this PR, as it will cause the performance regression.
If we just want to stop pushing down the non-deterministic expression without considering the regression, probably a fewer code needed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 5]
```diff
@@ -513,18 +513,28 @@ object SimplifyFilters extends Rule[LogicalPlan] {
...
 
-  private def replaceAlias(condition: Expression, sourceAliases: Map[Attribute, Expression]) = {
-    condition transform {
-      case a: AttributeReference => sourceAliases.getOrElse(a, a)
-    }
+      // We only push down filter if their overlapped expressions are all
+      // deterministic.
+      val hasNondeterministic = condition.collect {
+        case a: Attribute if aliasMap.contains(a) => aliasMap(a)
+      }.exists(_.find(!_.deterministic).isDefined)
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]:
```scala
_.find(!_.deterministic).isDefined
```
this is also just
```scala
_.exists(!_.deterministic)
```
right?

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: no, here we call `find` on a tree node, not a scala collection, and tree node don't provide `exists` function.

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: shall we add it?

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: i don't have a strong preference here

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 6]
```diff
@@ -216,11 +216,9 @@ object ProjectCollapsing extends Rule[LogicalPlan] {
 
       // We only collapse these two Projects if their overlapped expressions are all
       // deterministic.
-      val hasNondeterministic = projectList1.flatMap(_.collect {
-        case a: Attribute if aliasMap.contains(a) => aliasMap(a).child
-      }).exists(_.find(!_.deterministic).isDefined)
+      val canCollapse = projectList1.find(hasNondeterministic(_, aliasMap)).isEmpty
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: an unrelated but small change, we don't need to go through the whole project list, can stop once we find nondeterministic expressions.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 7]
```diff
@@ -234,6 +232,16 @@ object ProjectCollapsing extends Rule[LogicalPlan] {
         Project(substitutedProjection, child)
       }
   }
+
+  private def hasNondeterministic(
+    project: Expression,
+    sourceAliases: AttributeMap[Alias]) = {
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Nit: Use 4-space indentation please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 8]
```diff
@@ -216,11 +216,9 @@ object ProjectCollapsing extends Rule[LogicalPlan] {
 
       // We only collapse these two Projects if their overlapped expressions are all
       // deterministic.
-      val hasNondeterministic = projectList1.flatMap(_.collect {
-        case a: Attribute if aliasMap.contains(a) => aliasMap(a).child
-      }).exists(_.find(!_.deterministic).isDefined)
+      val noCollapse = projectList1.exists(hasNondeterministic(_, aliasMap))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: an unrelated but small change, we don't need to go through the whole project list, can stop once we find nondeterministic expressions.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 9]
```diff
@@ -234,6 +232,16 @@ object ProjectCollapsing extends Rule[LogicalPlan] {
         Project(substitutedProjection, child)
       }
   }
+
+  private def hasNondeterministic(
+      project: Expression,
+      sourceAliases: AttributeMap[Alias]) = {
+    project.exists {
+      case a: Attribute =>
+        sourceAliases.get(a).map(_.child.exists(!_.deterministic)).getOrElse(false)
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]:
```scala
        sourceAliases.get(a).exists(_.child.exists(!_.deterministic))
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 10]
```diff
@@ -512,20 +520,52 @@ object SimplifyFilters extends Rule[LogicalPlan] {
...
+  }
+
+  private def hasNondeterministic(
+      condition: Expression,
+      sourceAliases: AttributeMap[Expression]) = {
+    condition.exists {
+      case a: Attribute =>
+        sourceAliases.get(a).map(_.exists(!_.deterministic)).getOrElse(false)
+      case _ => false
+    }
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: @yhuai As we've once discussed offline, currently `Expression.deterministic` isn't recursively defined. Namely, `rand() + 1` is considered to be deterministic because `Add.deterministic` is `true`. My feeling is that making `Expression.deterministic` recursive looks more intuitive and simplifies things. For example, we can simply use `condition.deterministic` here without introducing all the `exists` calls, pattern matching, and the attribute map.

Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Is there any reason why `Expression.deterministic` wasn't designed to be recursive?

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: We have to define `deterministic` for all leaf expressions if we want to make it recursive...
Maybe it worth to do it now? cc @rxin

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: +1 on making deterministic describing the tree rather than the expression.

Might also make sense to have a trait called Nondeterministic that can be mixin into expressions, so we can easily find out what expressions are not deterministic.

**Comment by cloud-fan** [comment]
ping @yhuai

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 10 above)
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: @cloud-fan Actually we can make `Expression.deterministic` return `children.forall(_.deterministic)` by default. Then we don't need to define it for all leaf expressions.

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: already done in another PR :)

Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Which one?

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: https://github.com/apache/spark/pull/7525

Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Ah, it's already merged. No wonder I couldn't find it in the PR dashboard.

**Comment by liancheng** [comment]
@yhuai Sorry that I misunderstood your question at first. I think this PR is safe for normal filter push-down. Verified that partition pruning and Parquet filter push-down are both working properly.

**Comment by liancheng** [comment]
LGTM

**Comment by yhuai** [comment]
LGTM. I am merging it to master.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 11]
```diff
@@ -512,20 +512,50 @@ object SimplifyFilters extends Rule[LogicalPlan] {
...
+      // Create a map of Aliases to their values from the child projection.
+      // e.g., 'SELECT a + b AS c, d ...' produces Map(c -> a + b).
+      val aliasMap = AttributeMap(fields.collect {
+        case a: Alias => (a.toAttribute, a.child)
+      })
+
+      // Split the condition into small conditions by `And`, so that we can push down part of this
+      // condition without nondeterministic expressions.
+      val andConditions = splitConjunctivePredicates(condition)
+      val nondeterministicConditions = andConditions.filter(hasNondeterministic(_, aliasMap))
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Maybe use a `partition` at here will be better?
